File: sorting/heap.c
#include "helper_functions.h"

int parent(int ind)
{
    if(ind%2==0)
    {
        return (ind/2)-1;
    }
    return ind/2;
}


int left(int ind)
{
    return 2*ind+1;
}

int right(int ind)
{
    return 2*ind+2;
}
void downheap(int arr[], int ind, int size)
{
    int l = left(ind);
    int r = right(ind);
    int min = ind;

    if(l<size && arr[l]>arr[min])
    {
        min = l;
    }
    
    if(r<size && arr[r]>arr[min])
    {
        min = r;
    }

    if(ind!=min)
    {
        swap(&arr[min],&arr[ind]);
        downheap(arr, min, size);
    }

}

void heapify(int arr[], int size)
{
    for(int i=size/2-1  ;i>=0;i--)
    {
        downheap(arr  ,i, size);
    }
}

void heapsort(int arr[], int size)
{
    heapify(arr, size);
    for (int i=size-1; i>0 ; i--)
    {
        swap(&arr[0], &arr[i]);
            downheap(arr, 0 , i);

    }
    

}


int main()
{
    int size =10;
    int arr[size];

    generateRandomArray(arr, size, 0, 32);
      
    printf("Before sorting :\n");
    printArray(arr,size);

    heapsort(arr,size);
    printf("After sorting :\n");
    printArray(arr,size);
     if (isSorted(arr, size)) {
        printf("The array is sorted.\n");
    } else {
        printf("The array is not sorted.\n");
    }


    return 0;
}

File: Trees/Heap/heap.py
class Heap:
    def __init__(self):
        self.arr = []
    
    def swap(self, i, j):
        self.arr[i], self.arr[j] = self.arr[j], self.arr[i]
    
    def parent(self, index):
        return (index - 1) // 2
    
    def left(self, index):
        return (2 * index) + 1
    
    def right(self, index):
        return (2 * index) + 2
    
    def upheap(self, index):
        if(index ==0):
            return
        p = self.parent(index)
        if(self.arr[p]>self.arr[index]):
            self.swap(p, index)
            self.upheap(p)

    def downheap(self, index):
        min  = index
        left = self.left(index)
        right = self.right(index)
        if left < len(self.arr) and self.arr[left] < self.arr[min]:
            min = left
        if right < len(self.arr) and self.arr[right] < self.arr[min]:
            min = right
        if(min!=index):
            self.swap(min,index)
            self.downheap(min)
            

    def insert(self, data):
        self.arr.append(data)
        self.upheap(len(self.arr)-1)

    def remove(self):
        if(len(self.arr)==0):
            return  None
        top = self.arr[0]
        last = self.arr.pop()
        if(len(self.arr)!= 0):
            self.arr[0]= last   
            self.downheap(0)
        return top

    def heapsort(self):
        sort = []
        while len(self.arr) != 0:
            sort.append(self.remove())
        
        return sort



h = Heap()
h.insert(5)

h.insert(3)
h.insert(54)
h.insert(53)
h.insert(54)
print(h.heapsort())